storage/fcds/fcds.go

```diff
@@ -0,0 +1,358 @@
+	s = &Store{
+		shards:       make([]shard, shardCount),
+		meta:         metaStore,
+		free:         make([]bool, shardCount),
+		maxChunkSize: maxChunkSize,
+		quit:         make(chan struct{}),
+	}
+	for _, o := range opts {
+		o(s)
+	}
+	if err := os.MkdirAll(path, 0777); err != nil {
+		return nil, err
+	}
+	for i := byte(0); i < shardCount; i++ {
+		s.shards[i].f, err = os.OpenFile(filepath.Join(path, fmt.Sprintf("chunks-%v.db", i)), os.O_CREATE|os.O_RDWR, 0666)
+		if err != nil {
+			return nil, err
+		}
+		s.shards[i].mu = new(sync.Mutex)
+	}
+	return s, nil
+}
+
+// Get returns a chunk with data.
+func (s *Store) Get(addr chunk.Address) (ch chunk.Chunk, err error) {
+	if err := s.protect(); err != nil {
+		return nil, err
+	}
+	defer s.unprotect()
+
+	sh := s.shards[getShard(addr)]
+	sh.mu.Lock()
+	defer sh.mu.Unlock()
+
+	m, err := s.getMeta(addr)
+	if err != nil {
+		return nil, err
+	}
+	data := make([]byte, m.Size)
+	n, err := sh.f.ReadAt(data, m.Offset)
+	if err != nil && err != io.EOF {
+		return nil, err
+	}
+	if n != int(m.Size) {
+		return nil, fmt.Errorf("incomplete chunk data, read %v of %v", n, m.Size)
+	}
+	return chunk.NewChunk(addr, data), nil
+}
+
+// Has returns true if chunk is stored.
+func (s *Store) Has(addr chunk.Address) (yes bool, err error) {
+	if err := s.protect(); err != nil {
+		return false, err
+	}
+	defer s.unprotect()
+
+	mu := s.shards[getShard(addr)].mu
+	mu.Lock()
+	defer mu.Unlock()
+
+	_, err = s.getMeta(addr)
+	if err != nil {
+		if err == chunk.ErrChunkNotFound {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+// Put stores chunk data.
+func (s *Store) Put(ch chunk.Chunk) (err error) {
+	if err := s.protect(); err != nil {
+		return err
+	}
+	defer s.unprotect()
+
+	addr := ch.Address()
+	data := ch.Data()
+
+	section := make([]byte, s.maxChunkSize)
+	copy(section, data)
+
+	shard := getShard(addr)
+	sh := s.shards[shard]
+
+	sh.mu.Lock()
+	defer sh.mu.Unlock()
+
+	offset, reclaimed, err := s.getOffset(shard)
+	if err != nil {
+		return err
+	}
+
+	if offset < 0 {
+		offset, err = sh.f.Seek(0, io.SeekEnd)
+	} else {
+		_, err = sh.f.Seek(offset, io.SeekStart)
+	}
+	if err != nil {
+		return err
+	}
+
+	if _, err = sh.f.Write(section); err != nil {
+		return err
+	}
+	if reclaimed && s.freeCache != nil {
+		s.freeCache.remove(shard, offset)
+	}
+	return s.meta.Set(addr, shard, reclaimed, &Meta{
+		Size:   uint16(len(data)),
+		Offset: offset,
+	})
+}
+
+// getOffset returns an offset where chunk data can be written to
+// and a flag if the offset is reclaimed from a previously removed chunk.
+// If offset is less then 0, no free offsets are available.
+func (s *Store) getOffset(shard uint8) (offset int64, reclaimed bool, err error) {
+	if !s.shardHasFreeOffsets(shard) {
+		return -1, false, nil
+	}
+
+	offset = -1
+	if s.freeCache != nil {
+		offset = s.freeCache.get(shard)
+	}
+
+	if offset < 0 {
```

Comment:
Actually, no, the first `if` changes the `offset` value.

Reply:
yes, but why not just have all the logic in one block? you don't think it's more readable?

Reply:
Well, one block is handling the cache only, somehow I find that separation clear. Do you have a suggestion how would you change it?

Reply:
i was thinking that having it inside the last if block would make sense because it will always execute after the first `if` block has executed. it's really nitpicking, you can leave as is